mm2src/coins/nft_storage/sql_storage.rs

```diff
@@ -0,0 +1,885 @@
+}
+
+fn nft_from_row(row: &Row<'_>) -> Result<Nft, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn tx_history_from_row(row: &Row<'_>) -> Result<NftTransferHistory, SqlError> {
+    let json_string: String = row.get(0)?;
+    json::from_str(&json_string).map_err(|e| SqlError::FromSqlConversionFailure(0, Type::Text, Box::new(e)))
+}
+
+fn token_address_id_from_row(row: &Row<'_>) -> Result<NftTokenAddrId, SqlError> {
+    let token_address: String = row.get("token_address")?;
+    let token_id_str: String = row.get("token_id")?;
+    let token_id = BigDecimal::from_str(&token_id_str).map_err(|_| SqlError::from(FromSqlError::InvalidType))?;
+    Ok(NftTokenAddrId {
+        token_address,
+        token_id,
+    })
+}
+
+fn insert_nft_in_list_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_list_table_name(chain);
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            token_address, token_id, chain, amount, block_number, contract_type, details_json
+        ) VALUES (
+            ?1, ?2, ?3, ?4, ?5, ?6, ?7
+        );",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn insert_tx_in_history_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_tx_history_table_name(chain);
+    validate_table_name(&table_name)?;
+
+    let sql = format!(
+        "INSERT OR IGNORE INTO {} (
+            transaction_hash, chain, block_number, block_timestamp, contract_type,
+            token_address, token_id, status, amount, collection_name, image, token_name, details_json
+        ) VALUES (
+            ?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13
+        );",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn insert_chain_into_scanned_blocks_sql() -> MmResult<String, SqlError> {
+    let table_name = scanned_nft_blocks_table_name();
+    validate_table_name(&table_name)?;
+    Ok(format!("INSERT OR IGNORE INTO {} (chain) VALUES (?1);", table_name))
+}
+
+fn update_details_json_by_token_add_id_sql<F>(chain: &Chain, table_name_creator: F) -> MmResult<String, SqlError>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
+
+    validate_table_name(&table_name)?;
+    let sql = format!(
+        "UPDATE {} SET details_json = ?1 WHERE token_address = ?2 AND token_id = ?3;",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn update_meta_by_tx_hash_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_tx_history_table_name(chain);
+
+    validate_table_name(&table_name)?;
+    let sql = format!(
+        "UPDATE {} SET collection_name = ?1, image = ?2, token_name = ?3, details_json = ?4 WHERE transaction_hash = ?5;",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn update_nft_amount_sql<F>(chain: &Chain, table_name_creator: F) -> MmResult<String, SqlError>
```

Comment:
Ooh, I see now why `INSERT OR IGNORE` is used in `insert_nft_in_list_sql`, you update the values using `UPDATE` statement. It's up to you and what you see fit if you want to merge this with `insert_nft_in_list_sql` by using `INSERT OR REPLACE` then. I guess it's better to have different functions for updates. Why are we using `IGNORE` though, shouldn't we allow correct errors to be returned if we used `insert_nft_in_list_sql` for an entry that already exists?

Reply:
>  I guess it's better to have different functions for updates.

Yeah, I would like to divide the logic.

> Why are we using IGNORE though, shouldn't we allow correct errors to be returned if we used insert_nft_in_list_sql for an entry that already exists?

Yeah, now I see that its better to return the error, added INSERT INTO